### PR TITLE
Support migrations in sort.

### LINKED ISSUE #22: sort_tables should sort migrations
Migrations should be sorted by time by ``sort_tables``. See also #389.

### LINKED ISSUE #117: Sort should support sorting migrations with empty suffix
It should be possible to pass a set of tables with > 0 migrations to sort but where the bookmark says the chunk to sort is empty. This allows us to workaround issues where we have a sorted set of migrations that we want to pass to sort.

See also #22.

### PR DESCRIPTION
**Pull request by jeromekelleher**
Closes #22
Closes #117

- [x] Tests that fully cover new/changed functionality.
- [x] Documentation including tutorial content if appropriate.
- [x] Changelogs, if there are API changes.

(I've not done any docs because I can't see anywhere we've said this stuff wasn't supported)

### CODE REVIEW

**Comment by jeromekelleher** [comment]
@petrelharp I've added sort support for migrations here. What do you think of the sort key? Should we make the sort requirement for migrations stricter now? We'd need to update msprime to make sure things are output in this order, I think.

File: c/tskit/tables.c [hunk 1]
```diff
@@ -4463,6 +4472,28 @@ cmp_edge(const void *a, const void *b)
...
 }
 
+static int
+cmp_migration(const void *a, const void *b)
+{
+    const migration_sort_t *ca = (const migration_sort_t *) a;
+    const migration_sort_t *cb = (const migration_sort_t *) b;
+
+    int ret = (ca->time > cb->time) - (ca->time < cb->time);
+    /* If time values are equal, sort by the source population */
```
Comment (awohns) [inline c/tskit/tables.c:4482]: Perhaps worth noting the full sorting rules in `docs/data-model.rst`. Right now only the nondecreasing time value is listed.

**Review by awohns — APPROVED** [review]

File: c/tskit/tables.c [hunk 2]
```diff
@@ -4463,6 +4472,28 @@ cmp_edge(const void *a, const void *b)
...
+    if (ret == 0) {
+        ret = (ca->source > cb->source) - (ca->source < cb->source);
```
Comment (benjeffery) [inline c/tskit/tables.c:4484]: Would be nice to get these branches tested - I guess it means handcrafting some tests and that's a but painful?

**Review by benjeffery — APPROVED** [review]
LGTM, just one question on coverage.

**Comment by petrelharp** [comment]
I vote to not add more sortedness requirements if we don't have a good algorithmic reason to. Plus, it's not obvious to me that `source` and `dest` are the natural next things to sort on - I would have thought `node` would come next? Or maybe `left`? So, I'd say we should *just* sort on `time` here, and add some more requirements if it turns out to be helpful.

**Comment by jeromekelleher** [comment]
Well, we need to choose some keys to sort on to make things deterministic (which we definitely want, we had loads of problems in the past with different C implementations having slightly different sort orders). I chose the current ordering because that's what comes out of msprime. So, my vote would be to formalise the order that pops out naturally from msprime (starting [here](https://github.com/tskit-dev/msprime/blob/main/lib/msprime.c#L4276) - sort by (source, dest, left). Unless there's a good reason otherwise, we might as well go with an existing ordering.

Agreed on not adding any more sortedness requirements to the input, that seems like it's more trouble that it's worth.

**Comment by petrelharp** [comment]
I'm a bit confused still: by "deterministic" you mean specifying a total order, right? Don't we need some *more* keys for that, since you could have more than one migration event with the same `time, source, dest`?

**Comment by jeromekelleher** [comment]
> I'm a bit confused still: by "deterministic" you mean specifying a total order, right? Don't we need some more keys for that, since you could have more than one migration event with the same time, source, dest

Yeah, I mean a fully specified ordering, because the output of sort isn't deterministic otherwise. Ah, you're right ``(time, source, dest, left)`` isn't fully specified, we'd need ``(time, source, dest, left, node)`` because we could have multiple nodes migrating at once.

**Comment by petrelharp** [comment]
Sure, sounds good. I guess that sorting of mutations can be skipped with the bookmark if it's a performance problem.  So, `sort` will make migrations sorted on all these things, but "valid tree sequence requirements" will still only require sortedness on `time`, right?

**Comment by jeromekelleher** [comment]
Yep, agreed

**Comment by jeromekelleher** [comment]
We should probably try to get #1108 in first, since @petrelharp has had to do a lot of messing around with it already. Unless you'd prefer to get this in first Peter? The updates I need to do are fairly small.

**Comment by petrelharp** [comment]
I would love to not have to do more mucking around with #1108, yes.

**Comment by benjeffery** [comment]
Is this ready for me to have a look over?

**Comment by jeromekelleher** [comment]
No, not yet. Thanks for the ping though, will do ASAP. (Writing slides for a talk tomorrow atm!)

File: c/tskit/tables.c (hunk 1 above)
Comment (jeromekelleher) [reply c/tskit/tables.c:4482]: We decided to not put any further constraints on the input requirements. The extra keys are to make sure that the sort is always fully specified.

**Comment by jeromekelleher** [comment]
Ready for final review and merge here @benjeffery

**Review by benjeffery — APPROVED** [review]
Great stuff, annoyingly I can't find anything to comment on!

**Pull request merged**